background_scripts/background_script.js

```diff
@@ -0,0 +1,27 @@
+/*
+  A user pointed out that the storage.sync feature actually doesn't work at all.
+  This background script attempts to copy any old data stored in `storage.sync` 
+  over to `storage.local`, which will be used from now on.
+*/
+
+function onError(error) {
+  console.log(error);
+}
+
+function mergeSyncData(localData) {
+    if (Object.keys(localData).length == 0) {
+        return;
+    }  // Already moved, or new user.
+    else {
+        let gettingSync = browser.storage.sync.get();
+        // `.set()` failure has no backup plan
+        gettingSync.then(syncData => browser.storage.local.set(syncData)).catch(onError);
+    }
+}
+
+function switchToLocalStorage() {
+    let gettingLocal = browser.storage.local.get();
+    gettingLocal.then(mergeSyncData, onError);
+}
+
+browser.runtime.onInstalled(switchToLocalStorage);
```

Comment:
I see the following problem:
A user updates the extension. The user is NOT logged in with his user account(or maybe there is no internet connection after the update, or server is down). If the user logs into its account now the data wouldn't be migrated, because the onInstalled got called already.

the documentation states:
```
Fired when the extension is first installed, when the extension is updated to a new version, and when the browser is updated to a new version.
```
So there could be the possibility to get the migration done on the next browser update, but this wouldn't work because the 'mergeSyncData' wouldn't do anything because stuff could already be written to the local storage.

I do not have an idea how to prevent this. Is there a way to check if the user is logged in? I must admit i have no idea whats happening with data from users which never used the sync. I thought there would be fallback to the local storage but can't find anything inside the documentation.